File: backend/src/test/java/com/fitapp/training/application/FakeCustomExerciseRepository.java
package com.fitapp.training.application;

import com.fitapp.training.domain.CustomExercise;
import com.fitapp.training.domain.CustomExerciseRepository;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

/**
 * In-memory {@link CustomExerciseRepository} fake for application/adapter unit tests, no database.
 */
public class FakeCustomExerciseRepository implements CustomExerciseRepository {

  private final Map<UUID, CustomExercise> byId = new LinkedHashMap<>();

  @Override
  public Optional<CustomExercise> findById(UUID id) {
    return Optional.ofNullable(byId.get(id));
  }

  @Override
  public CustomExercise save(CustomExercise exercise) {
    byId.put(exercise.id(), exercise);
    return exercise;
  }

  @Override
  public void deleteById(UUID id) {
    byId.remove(id);
  }

  @Override
  public List<CustomExercise> findByUserIdUpdatedSince(UUID userId, Instant since, int limit) {
    return byId.values().stream()
        .filter(e -> e.userId().equals(userId) && e.updatedAt().isAfter(since))
        .sorted((a, b) -> a.updatedAt().compareTo(b.updatedAt()))
        .limit(limit)
        .toList();
  }

  @Override
  public void deleteByUserId(UUID userId) {
    byId.values().removeIf(e -> e.userId().equals(userId));
  }
}

File: apps/mobile/test/fakes/FakeCustomExerciseRepository.ts
import { ok } from "@/shared/domain/Result";
import type { Id } from "@/shared/domain/Id";
import type { CustomExercise } from "@/features/routines/domain/CustomExercise";
import type { CustomExerciseRepository } from "@/features/routines/application/ports";

/** Fake en memoria de `CustomExerciseRepository` (`features/routines/application/ports.ts`). */
export class FakeCustomExerciseRepository implements CustomExerciseRepository {
  private readonly byId = new Map<Id, CustomExercise>();
  readonly savedExercises: CustomExercise[] = [];

  constructor(seed: CustomExercise[] = []) {
    for (const exercise of seed) {
      this.byId.set(exercise.id, exercise);
    }
  }

  async save(exercise: CustomExercise) {
    this.byId.set(exercise.id, exercise);
    this.savedExercises.push(exercise);
    return ok(undefined);
  }

  async findById(id: Id) {
    return ok(this.byId.get(id) ?? null);
  }

  clearCalls = 0;

  async clear() {
    this.clearCalls += 1;
    this.byId.clear();
    return ok(undefined);
  }
}
